### PR TITLE
Proof of concept new compare function

### PR DESCRIPTION
**Pull request by richfitz**
A proof of concept automatic compiled compare function using the new odin support for describing compare and data.

~Note that this does not currently work with ode models it seems~

~Uses https://github.com/mrc-ide/odin/pull/294~

### CODE REVIEW

File: tests/testthat/test-compare.R [hunk 1]
```diff
@@ -258,3 +258,73 @@ test_that("rewrite compare for gpu complains if indenting is bad", {
...
+  d <- dust::dust_data(
+    data.frame(time = t,
+               observed = runif(length(t), 0, 100)))
+  expect_null(mod$compare_data())
+  mod$set_data(d)
+  expect_null(mod$compare_data())
+  y <- mod$run(t[[1]])
+  expect_equal(
+    mod$compare_data(),
+    dnorm(d[[1]][[2]]$observed, drop(y), 0.5, TRUE))
```
Comment (weshinsley) [inline tests/testthat/test-compare.R:314]: Note for self - the gaps in my maths/stats make it not quite intuitive why/what/how these two are equal - but they are, and it works and that's how the comparison succeeds.

**Review by weshinsley — APPROVED** [review]
OK - this looks good to me and I nearly understand it - one question on generate_dust.R:680, and some things I need to look at myself more to understand what we're trying to do.

File: tests/testthat/test-gpu.R [hunk 2]
```diff
@@ -253,3 +253,31 @@ test_that("Can include a pair of integer vectors in gpu shared memory", {
...
+    data.frame(time = t,
+               observed = rnorm(length(t), 0, sqrt(t)),
+               another = 0L))
+
+  mod <- gen$new(pars, 0, 10, seed = 42, gpu_config = 0L)
+
+  expect_null(mod$compare_data())
+  mod$set_data(d)
+  expect_null(mod$compare_data())
+  y <- mod$run(t[[1]])
```
Comment (r-ash) [inline tests/testthat/test-gpu.R:279]: Does this actually run on a GPU in this test?

**Review by r-ash — CHANGES_REQUESTED** [review]
Looks good to me, just the comment about a missing `stop`

**Review by weshinsley — CHANGES_REQUESTED** [review]
Just the package version to bump

File: tests/testthat/test-gpu.R (hunk 2 above)
Comment (r-ash) [reply tests/testthat/test-gpu.R:279]: Ok nice, thought so! Didn't think there were any github action runners with a GPU about

File: R/generate_dust.R [hunk 3]
```diff
@@ -656,6 +652,35 @@ dust_compare_info <- function(dat, rewrite) {
...
+               "compare_data",
+               args,
+               body)
+}
+
+
+generate_dust_compare_collect <- function(dat) {
+  collect <- names(which(vcapply(dat$equations, "[[", "type") == "compare"))
+  sprintf("return %s;", paste(collect, collapse = " + "))
+}
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/generate_dust.R:681]: I've added a test here to exercise this. But this just generates a sum over all the individual components

File: R/generate_dust_equation.R [hunk 4]
```diff
@@ -48,7 +49,12 @@ generate_dust_equation <- function(eq, dat, rewrite, gpu, mixed) {
     if (eq$lhs != eq$name && !(eq$lhs %in% eq$depends$variables)) {
       req <- setdiff(req, eq$lhs)
     }
-    stopifnot(all(req %in% names(dat$gpu$access)))
+    if (dat$features$has_data) {
+      names_data <- names(Filter(function(x) x$location == "data",
+                                 dat$data$elements))
+      req <- setdiff(req, names_data)
+    }
+    err <- setdiff(req, names(dat$gpu$access))
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/generate_dust_equation.R]: I've put the old assertion back in, as we don't expect this to fail with user input, just when we're refactoring

**Review by r-ash — APPROVED** [review]

**Review by weshinsley — APPROVED** [review]

**Pull request merged**